barretenberg/cpp/src/barretenberg/avm_fuzzer/harness/internal_call.fuzzer.cpp

```diff
@@ -0,0 +1,331 @@
+        .pc = context->get_pc(),
+        .internal_call_id = internal_call_stack_manager.get_call_id(),
+        .internal_call_return_id = internal_call_stack_manager.get_return_call_id(),
+        .next_internal_call_id = internal_call_stack_manager.get_next_call_id(),
+    };
+}
+
+void fuzz_internal_call(std::vector<ExecutionEvent>& ex_events,
+                        std::unique_ptr<ContextInterface>& context,
+                        InternalCallStackManagerInterface& internal_call_stack_manager,
+                        uint32_t loc)
+{
+    ExecutionEvent ex_event = { .wire_instruction = call_instr,
+                                .before_context_event = fill_context_event(context, internal_call_stack_manager) };
+    // Execution.execute pre-dispatch:
+    context->set_next_pc(context->get_pc() + static_cast<uint32_t>(call_instr.size_in_bytes()));
+
+    // Execution.internal_call(context, loc) - internal_call_stack_manager.push() emits the internal call stack
+    // event:
+    internal_call_stack_manager.push(context->get_pc(), context->get_next_pc());
+    context->set_next_pc(loc);
+
+    // Execution.execute post-dispatch:
+    context->set_pc(context->get_next_pc());
+
+    ex_event.after_context_event = fill_context_event(context, internal_call_stack_manager);
+    ex_events.push_back(ex_event);
+}
+
+void fuzz_internal_return(std::vector<ExecutionEvent>& ex_events,
+                          std::unique_ptr<ContextInterface>& context,
+                          InternalCallStackManagerInterface& internal_call_stack_manager)
+{
+    // TODO(MW): Case where we do not have a return i.e other exit condition?
+    ExecutionEvent ex_event = { .wire_instruction = ret_instr,
+                                .before_context_event = fill_context_event(context, internal_call_stack_manager) };
+    // Execution.execute pre-dispatch:
+    context->set_next_pc(context->get_pc() + static_cast<uint32_t>(ret_instr.size_in_bytes()));
+
+    // Execution.internal_return(context):
+    try {
+        auto next_pc = internal_call_stack_manager.pop();
+        context->set_next_pc(next_pc);
+    } catch (const std::exception& e) {
+        // Do post-dispatch error handling from Execution.execute
+        ex_event.error = ExecutionError::OPCODE_EXECUTION;
+        context->set_gas_used(context->get_gas_limit()); // Consume all gas.
+        context->halt();
+        // In Execution.execute, we do the above then continue without throwing. Here I want to re-throw to pass
+        // the opcode error out of the loop, so repeating the post-dispatch code below (TODO(MW): cleanup):
+        // Execution.execute post-dispatch:
+        context->set_pc(context->get_next_pc());
+        ex_event.after_context_event = fill_context_event(context, internal_call_stack_manager);
+        ex_events.push_back(ex_event);
+        // Re-throw
+        throw OpcodeExecutionException("Internal return failed: " + std::string(e.what()));
```

Comment:
TODO: I need to clean this up - I just copied and pasted what `Execution.execute()` does for this error since I wanted to rethrow back to the main fuzzing path (which means repeating code)